src/connstat.cpp

```diff
@@ -0,0 +1,187 @@
+#include "connstat.h"
+
+#include <netlink/netlink.h>
+#include <netlink/socket.h>
+#include <netlink/msg.h>
+#include <netlink/route/link.h>
+#include <net/if.h>
+
+namespace dht {
+namespace net {
+
+/* forward declarations */
+static inline uint32_t nl_mgrp(uint32_t group);
+static int nlcb(struct nl_msg*, void*);
+
+ConnectivityStatus::ConnectivityStatus()
+    : nlsk(nlsk_init())
+    , logger_(new dht::Logger)
+{
+    if (nlsk_setup(nlsk.get()) == FAILURE)
+        throw std::runtime_error("nlsk_setup() failed\n");
+
+    thrd_ = std::thread([this] () { nl_event_loop_thrd(nlsk.get()); });
+    thrd_.detach();
+}
+
+void
+ConnectivityStatus::setTopicListener(connstat_event_cb ucb, unsigned int topic)
+{
+    switch (topic) {
+    case NEWADDR:
+    case DELADDR:
+    case ADDR:
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        nl_socket_add_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        break;
+    default:
+        break;
+    }
+    topic_cbs[topic] = ucb;
+}
+
+void
+ConnectivityStatus::removeTopicListener(unsigned int topic)
+{
+    switch (topic) {
+    case NEWADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        topic_cbs[topic] = nullptr;
+        break;
+    case DELADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV4_IFADDR, RTNLGRP_NONE);
+        topic_cbs[topic] = nullptr;
+        break;
+    default:
+        break;
+    }
+}
+
+void
+ConnectivityStatus::executer(unsigned int topic)
+{
+    if (*(topic_cbs+topic))
+        (*(topic_cbs+topic))(topic);
+}
+
+int
+ConnectivityStatus::nlsk_setup(nl_sock* nlsk)
+{
+    /* TODO: return proper status */
+    int status = FAILURE;
+    /*
+     * Notifications do not use sequence numbers, disable sequence number
+     * checking.
+     */
+    nl_socket_disable_seq_check(nlsk);
+
+    /*
+     * Define a callback function, which will be called for each notification
+     * received
+     */
+    nl_socket_modify_cb(nlsk, NL_CB_VALID, NL_CB_CUSTOM, nlcb, (void *)this);
```

Comment:
Just for info: https://en.cppreference.com/w/cpp/language/explicit_cast

Reply:
Same for other cast

Reply:
good read, thanks.